### ISSUE TITLE
thoughts / questions / first impressions

### ISSUE THREAD

**Issue by mbriggs** [trailblazer/trailblazer#191]
You guys probably get this sort of thing a lot, feel free to close if this isn't a good channel for it.

I am at the point where I have read all the docs available, and done a few small features in TRB, so I am probably not grasping / misunderstanding some intentions here. My first impression is that this will take you much further then the "Rails Way", but I also have some concerns. Would love to get dev feedback if you get the chance, and hopefully this will be somewhat helpful.

1) This may be a misunderstanding to me, but it seems like you are advocating Operations as the sole way to express business logic all the way down. I think operations are an interesting framework for domain services, but that is one part of a well architected application. Operations feel like they should be an interface to your bounded contexts in some way.

2) The idea of "concepts" seems to be finer grained then a bounded context. I get that impression based on the guidance of scoping operations under model namespaces, rather then under the aggregate or the context. Is that intentional? I have found bounded contexts to be essential in managing rails growth.

3) I feel like there should be some sort of mechanism to specify the Operation "contract". By that I mean when I look at a method called "do_thing(foo, bar, baz)" I know what that method needs to work. With an interface like "Create.(params_hash, deps_hash)" that is a lot less clear. In a CRUD scenario, model params == operation params, so the `contract` makes the interface super clear. But in less trivial situations, you have to read through the entire operation to understand how the different pieces interact with the unstructured data.

4) In a general way, rails is extremely CRUD oriented, which I feel is probably the biggest problem. TRB seems like an orders of magnitude improvement on how to build things that way, but what if you are not building a CRUD oriented app? (that is a part of every app, but the bits that make it "interesting" are typically not CRUD)

I am glad this project exists, I am sorry if this comes off as overly critical. My intention in writing this is partially to see if I understand things correctly, and partially to share an early learning experience :)

**Comment by apotonick** [comment]
Hi Matt,

1. Operations are the pivotal element in TRB architectures. An endpoint (aka controller) instantly delegates to an operation which orchestrates all business logic. It's up to you how/what you do in an operation, e.g. how many levels of nesting, your own service objects, etc. TRB 2.1 will make nesting operations a first-class thing to do. 

    It's simply a new way of structuring code without any overhead. The fact that controllers call operations in their actions is more of a side-effect, OPs don't have to be public, only.

2. Concepts are an abstract structural idea, they are not limited to AR model names. We actually mention that many times in the guides. For example, a concept could also be `invoicing` or `pdf_export`. Also, you can nest concepts. It's simply a convention implemented via Ruby's namespacing.

3. The operations public interface is meant to be functional, hence `call`. The idea: I don't need to know your OPs details, I just want to use it. However, the step DSL makes it quite convenient to learn what an operation does *internally* and furthermore, in 2.1 you can render (or visualize) what an operation does in what order, etc. Not sure how that is related to `params`, though?

4. TRB has absolutely nothing to do with CRUD - where does this misunderstanding come from? Please help out here. The fact that we offer CRUD basics (e.g. `Contract::Validate`) is a sub set of what TRB does - we do all things business logic.

**Comment by mbriggs** [comment]
Hey nick,

One thing I am thinking is that the docs may be aimed at folks coming at this from being believers in the Rails Way who have run into a few problems. I worked on a very large rails app for 5 years or so, and been part of struggling to make rails architecture scale during most of that period, I'm 
 coming at this as a believer in DDD, and someone who lost their faith in rails architecture a _really_ long time ago. What I want from TRB is documentation / sugar that i can point my team at and say "do this", and expect if they follow those ideas they will end up with something that will stay maintainable.

 1. I'll elaborate a bit -- I see well architected rails as having a basically not-rails codebase, which rails "calls into" via some sort of high level facades. In that context, I see operations as being a really interesting way to act as "glue" between rails and not-rails, since by (my) definition, that is all high level coordination code. My impression is that TRB advocates your whole codebase to be expressed in the context of operations? From what you are saying that seems to be correct, but TRB doesnt get too pushy about it? Or maybe coordination code should be operations, but computation code should be expressed as other sorts of services unless they are trivial?

 2.  It is mentioned, but there are a fair amount of examples where you namespace around a model, and in my experience its pretty rare that a bounded context will have a single entity. I know it is very hard to talk about big architecture ideas in the context of small snippets of code, because using big code ideas on small code just over complicates things. There is also a decent amount of discussion of the mechanics of namespacing, I'm assuming thats because The Rails Way doesn't do that? The impression I got is "You _can_ do a not-a-model namespace, but most of the time you should namespace around models" which runs counter to my experience, where most of the time an appropriate namespace is around multiple models, and single model is more the trivial case.

 3. What I mean is if I want to call a method in ruby, I can jump to the method and see what its parameters are. In TRB, those requirements are spread through the implementation details of a pipeline, since each step is pulling its own requirements out of an unstructured hash. In a trivial case, thats not a big deal, but if you had something more complex, you would need to know what each step requires to know how to call the operation. I asked in gitter about this, and someone said that they will sometimes list the param structure in a comment on the top of the file. 

Now if you are doing a CRUD-y operation, you have contracts, which give a very clear list of requirements that are validated by the actual code. I was thinking that if I adopt TRB, I could probably implement a contract-like macro that uses dry-struct or something to type / validate / document incoming params. Since I was writing all this anyways, I thought I would bring it up :)

4. Again, it is due to how much "air time" is given to those concepts in docs / examples / code provided by the gems. Something I worry a lot about in ruby is when gems start saying "All of this repetitive code can be expressed in 1 line!" early on in that large project I mentioned earlier, I was part of a push to remove https://github.com/hcatlin/make_resourceful from our codebase. Rails is very CRUD oriented, a lot of attempts people have made over the years on improving rails have been CRUD oriented, so when something starts talking about making CRUD more DRY, I start getting worried. I think the best way to mitigate this (as well as #2) is to spend a bit more time talking about starting with something simple / small, and the path you take to change it as it grows, rather then just talking about code in the small. Rails works great in the small already, the problems you have are trying to make it scale up to the large.

**Comment by apotonick** [comment]
Great discussion! :pray: 

1. Actually, TRB is not meant to touch any Rails (except ActiveRecord, in case you're using that). It's exactly as you say: the application is mainly non-Rails, which should've been advocated from Rails 0.x but never was.

    Also, there are no best practices/conventions *yet* how to structure code we don't cover (we have policies, contracts, some model-related code, etc.) - you're not limited to our macros but can hook your own callable objects into the flow and even nest them. You can even call other objects/functions yourself, but then you might lose our beautiful features such as tracing (see 3.).

2. Let's face it: the TRB docs still suck and we need a better example application that shows "all this". I started working on a bookkeeping application we use at TRB internally which will act as an open-source example once it's presentable. The fact our docs don't show the cool stuff and the book is terribly outdated are extremely frustrating for me. :boom: 

3. In TRB 2.1, every "step" can have an incoming and outgoing "contract" (dry-struct) and even *typing*, so you always know what's missing. Also, you can visualize deeply nested OPs visually, trace running operations (also visibly) and our debugger will let you follow the flow with ins and outs. 

    I know that in a "pure" Ruby implementation, this is "simpler" (it actually isn't, when people use shit like inheritance and `super` 9 times in a row), but TRB changes that on purpose: we *want* engineers to think in flows and not in hacking together some Ruby. I understand what you mean, though. I was very skeptical when I introduced it but the great adoption proofed that the Ruby community is ready for a new thinking.

4. The "CRUD misunderstanding™" could be solved with the example apps. I've seen TRB used in many apps, though, where people use it for all kinds of business logic that are not typical CRUD. The new mission statement "The advanced business logic framework" should also help clarifying that, once the Ruby community learns what business logic *is* (usually takes 2-4 years, but quickly accelerating in the past years, which is motivating). Do you have any ideas how to promote that better?

To sum this up: from the outside, it may look as if we're happy with our docs and they're completed or something. No way, there are so many improvements planned and we're working our asses of to provide a better DX (developer experience, learnt that one yesterday haha). Trailblazer is still a young company but I hope to be able to hire someone for writing and structuring docs at some point.

**Comment by apotonick** [comment]
Oh, do you actually want to help out? I hear you're interested in the ins and outs contracts??? :stuck_out_tongue_closed_eyes: Happy to have a chat!

**Comment by n-rodriguez** [comment]
> Trailblazer is still a young company but I hope to be able to hire someone for writing and structuring docs at some point.

have you reached the point? :smile:

**Comment by apotonick** [comment]
Actually yes, I myself will spend the next couple of weeks documenting! :grimacing:

**Comment by fighella** [comment]
If I can help you Nick Sutterizer I will. #vivalagermany
